src/lib.rs

```diff
@@ -4,41 +4,132 @@
 
 //! Rust bindings to the Python interpreter.
 //!
-//! Look at [the guide](https://pyo3.rs/) for a detailed introduction.
+//! PyO3 can be used to write native Python modules and run Python code and modules from Rust.
 //!
-//! # Ownership and Lifetimes
+//! See [the guide](https://pyo3.rs/) for a detailed introduction.
 //!
-//! Because all Python objects potentially have multiple owners, the concept of
-//! Rust mutability does not apply to Python objects.  As a result, PyO3 allows
-//! mutating Python objects even if they are not stored in a mutable Rust
-//! variable.
+//! # PyO3's object types
 //!
-//! In Python, all objects are implicitly reference counted.  The Python
-//! interpreter uses a global interpreter lock (GIL) to ensure thread-safety.
-//! Thus, we use `struct Python<'py>` as a token to indicate that
-//! a function can assume that the GIL is held.  In Rust, we use different types
-//! to represent a reference to a Python object, depending on whether we know
-//! the GIL is held, and depending on whether we know the underlying type.  See
-//! [the guide](https://pyo3.rs/main/types.html) for an explanation of
-//! the different Python object types.
+//! PyO3 has several core types that you should familiarize yourself with:
 //!
-//! A `Python` instance is either obtained explicitly by acquiring the GIL,
-//! or implicitly by PyO3 when it generates the wrapper code for Rust functions
-//! and structs wrapped as Python functions and objects.
+//! ## The Python<'py> object
 //!
-//! # Error Handling
+//! Holding the [global interpreter lock](https://docs.python.org/3/glossary.html#term-global-interpreter-lock)
+//! (GIL) is modeled with the [`Python<'py>`](crate::Python) token.
+//! All APIs that require that the GIL is held require this token as proof
+//! that you really are holding the GIL. It can be explicitly acquired and
+//! is also implicitly acquired by PyO3 as it wraps Rust functions and structs
+//! into Python functions and objects.
 //!
-//! The vast majority of operations in this library will return `PyResult<...>`.
+//! ## The GIL-dependent types
+//!
+//! For example `&`[`PyAny`](crate::types::PyAny).
+//! These are only ever seen as references, with a lifetime that is only valid for as long
+//! as the GIL is held, which is why using them doesn't require a  [`Python<'py>`](crate::Python) token.
+//!  The underlying Python object, if mutable, can be mutated through any reference.
+//!
+//! See the [guide](https://pyo3.rs/main/types.html) for an explanation of the different Python object types.
+//!
+//! ## The GIL-independent types
+//!
+//! When wrapped in [`Py`]`<...>`, like with [`Py`]`<`[`PyAny`](crate::types::PyAny)`>` or [`Py`]`<SomePyClass>`, Python objects
+//! no longer have a limited lifetime which makes them easier to store in structs and pass between functions.
+//! However, you cannot do much with them without a
+//! [`Python<'py>`](crate::Python) token, for which you’d need to reacquire the GIL.
```

Comment:
Not mentioning `PyObject` here is intended, I assume? (It's probably for the best since it's supposed to be phased out anyway.)

Reply:
Yeah; imo it's pretty confusing to have both uses of `Py<PyAny>` and `PyObject` mixed through documentation, which is why I thought it best not to mention it.